File: clients/ts/nucl-parquet/src/relaxation.ts
/**
 * EADL atomic relaxation data (fluorescence X-rays and Auger electrons).
 *
 * After a photoelectric interaction creates a vacancy in an inner shell,
 * the atom relaxes via radiative (X-ray) or non-radiative (Auger) transitions.
 * This module provides the transition probabilities and energies needed to
 * sample the resulting secondary particles.
 */

import { parquetRead } from "hyparquet";
import { Transition, TransitionType } from "./types.js";

/** Load file as ArrayBuffer — Node.js implementation. */
async function loadFileAsArrayBuffer(path: string): Promise<ArrayBuffer> {
  const { readFile } = await import("fs/promises");
  const buffer = await readFile(path);
  return buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
}

/** Read a Parquet file and return rows. */
async function readParquetFile(path: string): Promise<Record<string, unknown>[][]> {
  const arrayBuffer = await loadFileAsArrayBuffer(path);
  const rows: Record<string, unknown>[][] = [];
  await parquetRead({
    file: {
      byteLength: arrayBuffer.byteLength,
      slice(start: number, end: number) {
        return arrayBuffer.slice(start, end);
      },
    },
    onComplete(data: Record<string, unknown>[][]) {
      rows.push(...data);
    },
  });
  return rows;
}

/** List .parquet files in a directory. */
async function listParquetFiles(dir: string): Promise<string[]> {
  const { readdir } = await import("fs/promises");
  const { join } = await import("path");
  const entries = await readdir(dir);
  return entries
    .filter((e) => e.endsWith(".parquet"))
    .map((e) => join(dir, e));
}

/** Check if a directory exists. */
async function dirExists(dir: string): Promise<boolean> {
  try {
    const { stat } = await import("fs/promises");
    const s = await stat(dir);
    return s.isDirectory();
  } catch {
    return false;
  }
}

/**
 * EADL atomic relaxation database.
 *
 * Use the static `open()` factory method to create an instance.
 *
 * @example
 * ```ts
 * const db = await RelaxationDb.open("path/to/nucl-parquet/meta");
 * const fy = db.fluorescenceYield(29, "K"); // ~0.44
 * ```
 */
export class RelaxationDb {
  private transitionMap: Map<number, Transition[]>;

  private constructor(transitionMap: Map<number, Transition[]>) {
    this.transitionMap = transitionMap;
  }

  /**
   * Load EADL data from the nucl-parquet `meta/` directory.
   *
   * Reads `meta/eadl/*.parquet`.
   */
  static async open(metaDir: string): Promise<RelaxationDb> {
    const { join } = await import("path");
    const dir = join(metaDir, "eadl");
    const transitionMap = new Map<number, Transition[]>();

    if (!(await dirExists(dir))) {
      return new RelaxationDb(transitionMap);
    }

    const files = await listParquetFiles(dir);
    for (const file of files) {
      const rows = await readParquetFile(file);

      let zVal: number | undefined;
      const transList: Transition[] = [];

      for (const rowGroup of rows) {
        for (const row of rowGroup) {
          const z = row["Z"] as number;
          if (zVal === undefined) zVal = z;

          transList.push({
            vacancyShell: row["vacancy_shell"] as string,
            fillingShell: row["filling_shell"] as string,
            transitionType:
              (row["transition_type"] as string) === "radiative"
                ? TransitionType.Radiative
                : TransitionType.Auger,
            energyKeV: row["energy_keV"] as number,
            probability: row["probability"] as number,
            edgeKeV: row["edge_keV"] as number,
          });
        }
      }

      if (zVal !== undefined) {
        // Sort by shell, then probability descending
        transList.sort((a, b) => {
          const shellCmp = a.vacancyShell.localeCompare(b.vacancyShell);
          if (shellCmp !== 0) return shellCmp;
          return b.probability - a.probability;
        });
        transitionMap.set(zVal, transList);
      }
    }

    return new RelaxationDb(transitionMap);
  }

  /** Get all transitions for element Z. */
  transitions(z: number): Transition[] {
    return this.transitionMap.get(z) ?? [];
  }

  /** Get transitions for a specific vacancy shell (e.g., "K"). */
  shellTransitions(z: number, shell: string): Transition[] {
    return this.transitions(z).filter((t) => t.vacancyShell === shell);
  }

  /** Get only radiative (X-ray) transitions for element Z and shell. */
  radiativeTransitions(z: number, shell: string): Transition[] {
    return this.transitions(z).filter(
      (t) => t.vacancyShell === shell && t.transitionType === TransitionType.Radiative,
    );
  }

  /** Fluorescence yield for a shell = sum of radiative transition probabilities. */
  fluorescenceYield(z: number, shell: string): number {
    return this.transitions(z)
      .filter((t) => t.vacancyShell === shell && t.transitionType === TransitionType.Radiative)
      .reduce((sum, t) => sum + t.probability, 0);
  }

  /** Check if data is loaded for element Z. */
  hasElement(z: number): boolean {
    return this.transitionMap.has(z);
  }
}

File: clients/rs/nucl-parquet/src/relaxation.rs
//! EADL atomic relaxation data (fluorescence X-rays and Auger electrons).
//!
//! After a photoelectric interaction creates a vacancy in an inner shell,
//! the atom relaxes via radiative (X-ray) or non-radiative (Auger) transitions.
//! This module provides the transition probabilities and energies needed to
//! sample the resulting secondary particles.

use std::collections::HashMap;
use std::fs;
use std::path::Path;

use arrow::array::{Float64Array, Int32Array, StringArray};
use parquet::arrow::arrow_reader::ParquetRecordBatchReaderBuilder;

/// Type of atomic relaxation transition.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TransitionType {
    /// Radiative transition — emits a characteristic X-ray photon.
    Radiative,
    /// Auger transition — emits an electron.
    Auger,
}

/// A single atomic relaxation transition.
#[derive(Debug, Clone)]
pub struct Transition {
    /// Shell where the vacancy was created (e.g., "K", "L1").
    pub vacancy_shell: String,
    /// Shell that fills the vacancy (e.g., "L3").
    pub filling_shell: String,
    /// Radiative (X-ray) or Auger (electron).
    pub transition_type: TransitionType,
    /// Transition energy in keV.
    pub energy_kev: f64,
    /// Transition probability (fractional, sums to ~1 per shell).
    pub probability: f64,
    /// Binding energy of the vacancy shell in keV.
    pub edge_kev: f64,
}

/// EADL atomic relaxation database.
///
/// Thread-safe: `Send + Sync`.
pub struct RelaxationDb {
    /// Z -> list of transitions (sorted by vacancy_shell, then probability desc)
    transitions: HashMap<u8, Vec<Transition>>,
}

impl RelaxationDb {
    /// Load EADL data from the nucl-parquet `meta/` directory.
    ///
    /// Reads `meta/eadl/*.parquet`.
    pub fn open(meta_dir: impl AsRef<Path>) -> crate::Result<Self> {
        let dir = meta_dir.as_ref().join("eadl");
        let mut transitions: HashMap<u8, Vec<Transition>> = HashMap::new();

        if !dir.exists() {
            return Ok(Self { transitions });
        }

        for entry in fs::read_dir(&dir)? {
            let entry = entry?;
            let path = entry.path();
            if path.extension().and_then(|e| e.to_str()) != Some("parquet") {
                continue;
            }

            let file = fs::File::open(&path)?;
            let reader = ParquetRecordBatchReaderBuilder::try_new(file)?.build()?;

            let mut z_val: Option<u8> = None;
            let mut trans_list = Vec::new();

            for batch in reader {
                let batch = batch?;

                let z_col = batch
                    .column_by_name("Z")
                    .and_then(|c| c.as_any().downcast_ref::<Int32Array>());
                let vacancy_col = batch
                    .column_by_name("vacancy_shell")
                    .and_then(|c| c.as_any().downcast_ref::<StringArray>());
                let filling_col = batch
                    .column_by_name("filling_shell")
                    .and_then(|c| c.as_any().downcast_ref::<StringArray>());
                let type_col = batch
                    .column_by_name("transition_type")
                    .and_then(|c| c.as_any().downcast_ref::<StringArray>());
                let energy_col = batch
                    .column_by_name("energy_keV")
                    .and_then(|c| c.as_any().downcast_ref::<Float64Array>());
                let prob_col = batch
                    .column_by_name("probability")
                    .and_then(|c| c.as_any().downcast_ref::<Float64Array>());
                let edge_col = batch
                    .column_by_name("edge_keV")
                    .and_then(|c| c.as_any().downcast_ref::<Float64Array>());

                if let (Some(z), Some(vac), Some(fill), Some(tt), Some(e), Some(p), Some(edge)) = (
                    z_col,
                    vacancy_col,
                    filling_col,
                    type_col,
                    energy_col,
                    prob_col,
                    edge_col,
                ) {
                    for i in 0..batch.num_rows() {
                        if z_val.is_none() {
                            z_val = Some(z.value(i) as u8);
                        }
                        trans_list.push(Transition {
                            vacancy_shell: vac.value(i).to_string(),
                            filling_shell: fill.value(i).to_string(),
                            transition_type: match tt.value(i) {
                                "radiative" => TransitionType::Radiative,
                                _ => TransitionType::Auger,
                            },
                            energy_kev: e.value(i),
                            probability: p.value(i),
                            edge_kev: edge.value(i),
                        });
                    }
                }
            }

            if let Some(z) = z_val {
                // Sort by shell, then probability descending
                trans_list.sort_by(|a, b| {
                    a.vacancy_shell
                        .cmp(&b.vacancy_shell)
                        .then(b.probability.partial_cmp(&a.probability).unwrap())
                });
                transitions.insert(z, trans_list);
            }
        }

        Ok(Self { transitions })
    }

    /// Get all transitions for element Z.
    pub fn transitions(&self, z: u8) -> &[Transition] {
        self.transitions
            .get(&z)
            .map(|v| v.as_slice())
            .unwrap_or(&[])
    }

    /// Get transitions for a specific vacancy shell (e.g., "K").
    pub fn shell_transitions(&self, z: u8, shell: &str) -> Vec<&Transition> {
        self.transitions(z)
            .iter()
            .filter(|t| t.vacancy_shell == shell)
            .collect()
    }

    /// Get only radiative (X-ray) transitions for element Z and shell.
    pub fn radiative_transitions(&self, z: u8, shell: &str) -> Vec<&Transition> {
        self.transitions(z)
            .iter()
            .filter(|t| t.vacancy_shell == shell && t.transition_type == TransitionType::Radiative)
            .collect()
    }

    /// Fluorescence yield for a shell = sum of radiative transition probabilities.
    pub fn fluorescence_yield(&self, z: u8, shell: &str) -> f64 {
        self.transitions(z)
            .iter()
            .filter(|t| t.vacancy_shell == shell && t.transition_type == TransitionType::Radiative)
            .map(|t| t.probability)
            .sum()
    }

    /// Check if data is loaded for element Z.
    pub fn has_element(&self, z: u8) -> bool {
        self.transitions.contains_key(&z)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[ignore = "requires nucl-parquet data files"]
    fn load_and_query_cu() {
        let db = RelaxationDb::open("../../meta").unwrap();
        assert!(db.has_element(29));

        let k_trans = db.shell_transitions(29, "K");
        assert!(!k_trans.is_empty());

        // K fluorescence yield for Cu should be ~0.44
        let fy = db.fluorescence_yield(29, "K");
        assert!(fy > 0.3 && fy < 0.6, "Cu K fluorescence yield: {fy}");

        // K-alpha line should be ~8 keV
        let radiative = db.radiative_transitions(29, "K");
        let strongest = radiative
            .iter()
            .max_by(|a, b| a.probability.partial_cmp(&b.probability).unwrap());
        assert!(strongest.is_some());
        let ka = strongest.unwrap();
        assert!(
            ka.energy_kev > 7.0 && ka.energy_kev < 9.0,
            "Cu K-alpha: {} keV",
            ka.energy_kev
        );
    }
}
